File: 二叉树/二叉搜索树迭代器.go
package 二叉树

type BSTIterator struct {
	ptr  int
	nums []int
}

func Constructor(root *TreeNode) BSTIterator {
	var xx []int
	xx = append(xx, -114514)
	var inorder func(root *TreeNode)
	inorder = func(root *TreeNode) {
		if root == nil {
			return
		}
		inorder(root.Left)
		xx = append(xx, root.Val)
		inorder(root.Right)
	}
	inorder(root)
	return BSTIterator{0, xx}
}

func (this *BSTIterator) Next() int {
	this.ptr++
	return this.nums[this.ptr]
}

func (this *BSTIterator) HasNext() bool {
	if this.ptr == len(this.nums)-1 {
		return false
	} else {
		return true
	}
}

File: 二叉树/二叉搜索树迭代器.java
package 二叉树;

import java.util.ArrayList;

public class 二叉搜索树迭代器 {
}

//剑指offer
class BSTIterator {
    ArrayList<Integer> list;
    int ptr = 0;

    public BSTIterator(TreeNode root) {
        list = new ArrayList<Integer>();
        list.add(Integer.MIN_VALUE);
        dfs(root);
    }

    public int next() {
        ptr++;
        return list.get(ptr);
    }

    public boolean hasNext() {
        if (ptr+1 <= list.size()-1)  return true;
        return false;
    }

    public void dfs(TreeNode root) {
        if (root == null) return;
        dfs(root.left);
        list.add(root.val);
        dfs(root.right);
    }
}